tests/tape/test_tape_measure.py

```diff
@@ -501,3 +512,360 @@ def func():
+        """Test that the density matrix is of correct size and dtype for a
+        trivial circuit"""
+
+        dev = qml.device("default.qubit", wires=wires)
+
+        @qnode(dev)
+        def circuit():
+            return density_matrix([0])
+
+        state_val = circuit()
+
+        assert state_val.shape == (2, 2)
+        assert state_val.dtype == np.complex128
+
+    @pytest.mark.parametrize("wires", range(2, 5))
+    def test_density_matrix_shape_and_dtype_mixed(self, wires):
+        """Test that the density matrix is of correct size and dtype for a
+        trivial circuit"""
+
+        dev = qml.device("default.mixed", wires=wires)
+
+        @qnode(dev)
+        def circuit():
+            return density_matrix([0])
+
+        state_val = circuit()
+
+        assert state_val.shape == (2, 2)
+        assert state_val.dtype == np.complex128
+
+    def test_return_type_is_state_qubit(self):
+        """Test that the return type of the observable is State"""
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qnode(dev)
+        def func():
+            qml.Hadamard(0)
+            return density_matrix(0)
+
+        func()
+        obs = func.qtape.observables
+        assert len(obs) == 1
+        assert obs[0].return_type is State
+
+    def test_return_type_is_state_mixed(self):
+        """Test that the return type of the observable is State"""
+
+        dev = qml.device("default.mixed", wires=2)
+
+        @qnode(dev)
+        def func():
+            qml.Hadamard(0)
+            return density_matrix(0)
+
+        func()
+        obs = func.qtape.observables
+        assert len(obs) == 1
+        assert obs[0].return_type is State
+
+    def test_correct_density_matrix_product_state_first_qubit(self):
+        """Test that the correct density matrix  is returned when
+        trace out a product state"""
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qnode(dev)
+        def func():
+            qml.Hadamard(wires=1)
+            qml.PauliY(wires=0)
+            return density_matrix(0)
+
+        density_first = func()
+
+        assert np.allclose(
+            np.array([[0.0 + 0.0j, 0.0 + 0.0j], [0.0 + 0.0j, 1.0 + 0.0j]]), density_first
+        )
+
+    def test_correct_density_matrix_product_state_first_mixed(self):
+        """Test that the correct density matrix  is returned when
+        trace out a product state"""
+
+        dev = qml.device("default.mixed", wires=2)
+
+        @qnode(dev)
+        def func():
+            qml.Hadamard(wires=1)
+            qml.PauliY(wires=0)
+            return density_matrix(0)
+
+        density_first = func()
+
+        assert np.allclose(
+            np.array([[0.0 + 0.0j, 0.0 + 0.0j], [0.0 + 0.0j, 1.0 + 0.0j]]), density_first
+        )
+
+    def test_correct_density_matrix_product_state_second_qubit(self):
+        """Test that the correct density matrix  is returned when
+        trace out a product state"""
```

Comment:
```suggestion
        tracing out a product state"""
```

Reply:
This docstring is the same as the previous test - would be good to expand to detail how/why this test differs